feat(adr): addition of north-south messaging ADR (#656)
edgexfoundry/edgex-docs: docs_src/design/adr/0023-North-South-Messaging.md

In a future release of EdgeX, there is a desire to allow 3rd party systems to make requests of the southside via message bus.  Specifically, a 3rd party system will send a command request via message to the command service via the allowed message bus implementations (which could be MQTT or Redis Pub/Sub today).  The command service would then relay the message request via message bus to the managing device service.  (The command service would also perform any translations on the request as it does for REST requests today.) The device service would use the message to trigger action on the device/sensor as it does when it receives a REST request today and respond via message bus back to the command service.  In turn, the command service would respond to the 3rd party system via message bus.

### Design

> The command service requires the means to subscribe to the EdgeX message bus.  It would use the messaging client (go-mod-messaging) to create a new MessageClient, connect to the message bus, and subscribe to a configured message topic.

Author: I think it could be either.  An adopter may choose to use their own message bus or use the EdgeX provided one - depending on whether they want to originate requests through their bus or ours.  I changed the document to reflect it could be either and should be configurable.

Likewise, the device services must use the messaging client (go-mod-messaging for Go services and C alternative for C services) to create a new MessageClient, connect to the message bus and subscribe to configured message topics to receive actuation messages from the command service.

Because any actuation request may require a response on the part of EdgeX, the command service must provide the means to receive an optional response message from the device services and publish a response back to the 3rd party caller via message bus as well.

The command message request messages should contain a correlation identifier which would be added to any response so that the 3rd party system would know to associate the responses to the original request.

The device service message command subscription, response publishing and general handling should be implemented in the SDKs versus individually in the device services.

![image](command-msg.png)

### Message Structure

The messages to the command service and those relayed to the device service would mimic their REST alternatives.  The difference is that the HTTP REST requests

> - have a header which contain the correlation id. 

Author: My fault on this Cloud - correlation id would be in the header and I should have had the requestId in the body.  Instead I had correlation id in the body.  Changed accordingly.

- have a path that contains the target device and command name

- are made via HTTP GET or PUT methods signaling get or set actions on the device.

The message bus and the messages on the bus need this additional information. To address this, the command requests must be embedded in a “wrapper message” structure that would contain the additional elements that would be present in the HTTP request messages.

![image](command-msg-structure.png)

Specifically, the message structure would look something like the following:

``` json
{
“method” : “PUT”

> “name”: “sensor01”

Reviewer: rename to `deviceName` would be explicit

“command”:”command01”
“correlation-id”: “14a42ea6-c394-41c3-8bcd-a29b9f5e6835”

> “request”:

Author: fixed

        "AHU-TargetBand": "4.0",
        "AHU-TargetHumidity": {
        "Accuracy": "0.2-0.3% RH",
        "Value": 59
    }
}
```

Response messages would be similarly wrapped so as to preserve the correlation id - if nothing else.

``` json
{
“correlation-id”: “14a42ea6-c394-41c3-8bcd-a29b9f5e6835”
“response”:
    {
        "apiVersion": "v2",
        "requestId": "e6e8a2f4-eb14-4649-9e2b-175247911369",
        "statusCode": 0,
        "message": "string"
    }
}
```

### Configuration

The configuration or the command service must provide the details in order to subscribe to command messages from any 3rd party system and publish to a response topic.  The subscription configuration would be similar to how core data is configured to subscribe to messages from any device service and publish to topics to get messages to application services.  However, the command service will actually need to have two sets of subscription and two sets of publishing topics.

The command service would subscribe to any command topic (“edgex/request/command/#”) to receive requests from the 3rd party system.  It must then relay these messages (via publish) to the device service actuation topics.

> As each device service will have a topic assigned to it, the device service name and device name will be added to the destination topic (“edgex/request/actuate/<device-service>/<device-name>”) when command relays the actuation request messages to device services.

Reviewer: The character `“` causes some problem when displaying the content, so please try to use ` instead

The command service must then subscribe to another topic (“edgex/response/actuate/#”) to receive actuation responses from the device services.  It then relays those responses on to the 3rd party apps by publishing those responses to a response topic (“edgex/response/command”).

![image](command-service-topics.png)

``` toml
[MessageQueue]
Protocol = "redis"
Host = "localhost"
Port = 6379
Type = "redis"
AuthMode = "usernamepassword"  # required for redis messagebus (secure or insecure).
SecretName = "redisdb"

> SubscribeResponseActateTopic = “edgex/response/actuate/#”

Reviewer: Typo: Actuate

PublishActuateTopicPrefix = "edgex/actuate/ " # /<device-service>/<device-name> will be added to this Publish Topic prefix
SubscribeRequestCommandTopic = "edgex/request/command/#"
PublishCommandResponseTopic = “edgex/response/command”
```

Device services must similarly be configured for subscribing to actuation commands and responding to them by publishing to the response topic.

``` toml
[MessageQueue]
Protocol = "redis"
Host = "localhost"
Port = 6379
Type = "redis"
AuthMode = "usernamepassword"  # required for redis messagebus (secure or insecure).
SecretName = "redisdb"

> SubscribeRequestActateTopic = “edgex/request/actuate/<device-service>/#”

Reviewer: Typo: Actuate

PublishActuateResponseTopic = “edgex/response/actuate/<device-service>/”
```

Questions:

> -  Would we need multiple 3rd party response topics?  Or would one setup separate command services for each 3rd party?

Reviewer 2: Seems to me like it would work without considering what third party requested. Also it seems to me that one command service would suffice to handle all commands for all device services.

> -  Would we need multiple 3rd party response topics?  Or would one setup separate command services for each 3rd party?

Reviewer: The question might be that whether command service could subscribe the external message broker from 3rd party.  If it subscribes the EdgeX message bus (internal) only, we need additional services to relay the data to EdgeX message bus from external.
Dynamic configuration is not a user friendly approach. We might create additional APIs for Adding/Updating/Deleting/Query the external subscription and store them to the RedisDB.

Author: Added this to the questions section as it should be something discussed with the community.

> - Do we need separate topics for all the devices or would one on the device service suffice?

Reviewer 2: We definitely want to handle the case where a device service is managing two or more identical devices (so, having the same resources). Then it must be that the device is indicated by name.

> - Do we need separate topics for all the devices or would one on the device service suffice?

Reviewer: You have defined the deviceName in the payload, so one topic should be sufficient for Device Service- `edgex/request/actuate/<device-service>`

So, in summary, data can be collected from a sensor or device and be sent from the southside to the northside entirely using message bus technology when desired.

!!! Note
    The message bus may also be implemented in Zero MQ but with some limitations.  As Zero MQ is being deprecated in EdgeX, it will not be considered further in this ADR.

Communications from a 3rd party system (enterprise application, cloud application, etc.) to EdgeX in order to acuate a device or get the latest information from a sensor is accomplished via REST.  The 3rd party system makes a REST call of the command service which then relays a request to a device service also using REST.  There is no built in means to make a message-based request of EdgeX or the devices/sensors it manages.

> In a future release of EdgeX, there is a desire to allow 3rd party systems to make requests of the southside via message bus.  Specifically, a 3rd party system will send a command request via message to the command service via the allowed message bus implementations (which could be MQTT or Redis Pub/Sub today).  The command service would then relay the message request via message bus to the managing device service.  (The command service would also perform any translations on the request as it does for REST requests today.) The device service would use the message to trigger action on the device/sensor as it does when it receives a REST request today and respond via message bus back to the command service.  In turn, the command service would respond to the 3rd party system via message bus.

Author: @lenny-intel - it this true?  If I was writing an application in Go, then I might choose to use our go-mod.  But in another language, I just have to make sure I understand the message structures (and return information) and deal with that on my own.  Take our C device services - they aren't using our go-mod but still communicate with core data or the app services.  Obviously, we have incorporated that into our C SDK, but the model is there for a 3rd party to do the work independent of any go-mod.

> In a future release of EdgeX, there is a desire to allow 3rd party systems to make requests of the southside via message bus.  Specifically, a 3rd party system will send a command request via message to the command service via the allowed message bus implementations (which could be MQTT or Redis Pub/Sub today).  The command service would then relay the message request via message bus to the managing device service.  (The command service would also perform any translations on the request as it does for REST requests today.) The device service would use the message to trigger action on the device/sensor as it does when it receives a REST request today and respond via message bus back to the command service.  In turn, the command service would respond to the 3rd party system via message bus.

Reviewer 3: @jpwhitemn , correct "using our go-mod" is more abstract here, i.e. recreate it for other languages. The bigger issue is our EdgeX message bus is meant to be internal to EdgeX as far as the edgex box is concerned. I.e eKuiper connects to it, but is run as part of the EdgeX stack. But should we be allowing, off box, 3rd party systems to connect to our internal MessageBus?

> In a future release of EdgeX, there is a desire to allow 3rd party systems to make requests of the southside via message bus.  Specifically, a 3rd party system will send a command request via message to the command service via the allowed message bus implementations (which could be MQTT or Redis Pub/Sub today).  The command service would then relay the message request via message bus to the managing device service.  (The command service would also perform any translations on the request as it does for REST requests today.) The device service would use the message to trigger action on the device/sensor as it does when it receives a REST request today and respond via message bus back to the command service.  In turn, the command service would respond to the 3rd party system via message bus.

Reviewer 3: @tonyespy , @jpwhitemn , IMO, we still need the Command Service to be the External MQTT connection which then publishes the command request to the specific Device service for the requested Device/Command. As with REST the Command service knows with Device service to route to publish to based on the Device specified in the request.

Example:
   External topic received is : `edgex/command/request/<device>/<command-name>/...`
   Command Service forwards the command on the internal Edgex MessageBus with the topic: `edgex/command/request/<device-service>/<device>/<command-name>/...`

Alternatively we could have the Device Services subscribe to `edgex/command/request/<device>/#` for each device when they are added. DS would also have to unsubscribe when devices are removed. This would allow any EdgeX service to send a command via the internal EdgeX MessageBus using the topic `edgex/command/request/<device>/<command-name>/...`

> In a future release of EdgeX, there is a desire to allow 3rd party systems to make requests of the southside via message bus.  Specifically, a 3rd party system will send a command request via message to the command service via the allowed message bus implementations (which could be MQTT or Redis Pub/Sub today).  The command service would then relay the message request via message bus to the managing device service.  (The command service would also perform any translations on the request as it does for REST requests today.) The device service would use the message to trigger action on the device/sensor as it does when it receives a REST request today and respond via message bus back to the command service.  In turn, the command service would respond to the 3rd party system via message bus.

Reviewer 4: > Also as others (@cloudxxx8, @lenny-intel) have pointed out, for this to work we'd need to grant these 3rd party applications access to our message bus, which has already been a bit problematic with Kuiper. Since our default message bus impl is Redis pub/sub, this means we would need some kind of mechanism to securely distribute the Redis credentials (e.g. user/password) to these external applications.
> 
> It also should be noted that this approach would completely bypass the API Gateway.
> 
> If on the other hand, we secure our MQTT message bus implementation, authentication/authorization of external clients could be handled via client TLS client authentication, which plays into our discussions about Spiffe/Spire (see the [Delayed Services ADR](https://github.com/edgexfoundry/edgex-docs/blob/main/docs_src/design/adr/security/0020-spiffe.md) and/or [Microservices Auth ADR](https://github.com/edgexfoundry/edgex-docs/pull/659)).
> 
> Moving to a model where we instead use an external MQTT broker for this purpose means we've pushed all responsibility for securing communications to EdgeX to the entity operating this broker.

From a security perspective, there are two separate problems:

1. Granting parties the authorization to talk on the message bus.  Pretty much every MQTT broker has a way to do this, where the most common are username/password and certificate-based mutual-auth TLS.  SPIFFE/SPIRE is a good way to distribute such certificates if one can manage the requirements for certificate rotation on a continual basis.  Otherwise, username/password will probably end up being used (with server-side TLS if the broker is off-box)
2. Enabling the receiving microservice to positively identity the sending microservice.  It is an uncommon message broker feature for the message broker to pass metadata as to the sender of a message.  Usually, however, we have to roll our own, such as putting an access_token field in the message body that has a JWT, for example.

> In a future release of EdgeX, there is a desire to allow 3rd party systems to make requests of the southside via message bus.  Specifically, a 3rd party system will send a command request via message to the command service via the allowed message bus implementations (which could be MQTT or Redis Pub/Sub today).  The command service would then relay the message request via message bus to the managing device service.  (The command service would also perform any translations on the request as it does for REST requests today.) The device service would use the message to trigger action on the device/sensor as it does when it receives a REST request today and respond via message bus back to the command service.  In turn, the command service would respond to the 3rd party system via message bus.

Author: Added "Alert" note in this section that access to 3rd party systems is something to be aware of and consider, but left to the adopter as a decision for now.  Issue of internal vs external message bus is address per other comments later in the doc.

> In a future release of EdgeX, there is a desire to allow 3rd party systems to make requests of the southside via message bus.  Specifically, a 3rd party system will send a command request via message to the command service via the allowed message bus implementations (which could be MQTT or Redis Pub/Sub today).  The command service would then relay the message request via message bus to the managing device service.  (The command service would also perform any translations on the request as it does for REST requests today.) The device service would use the message to trigger action on the device/sensor as it does when it receives a REST request today and respond via message bus back to the command service.  In turn, the command service would respond to the 3rd party system via message bus.

Reviewer 4: > The command service would then relay the message request via message bus to the managing device service.

Any protocol translation or relay service (REST-MQTT, MQTT-REST, REST-REST, MQTT-MQTT) should pass along authentication tokens.

Author: noted.

### Design

> The command service requires the means to subscribe to the EdgeX or external (as a 3rd party may want to use their message bus to send commands) message bus.  It would use the messaging client (go-mod-messaging) to create a new MessageClient, connect to a message bus, and subscribe to a configured message topic.

Reviewer 5: As future work, does it make sense to extend the app-service-confurable service to not just act as exporter but also to import data (e.g. via an mqtt-import profile), internally using the proposed MessageClient or equivalent?

Author: This would require the app service to have more knowledge about device services, devices, etc. than it does today.  Separation of concerns allows the app service to be focused on data north and to data consumers like a rules engine.  Command service has always served as the proxy to the device services.

The command service would also publish messages to device service topics using the go-mod-messaging MessageClient.  It would relay any command request to a designated topic for the device service which would be subscribed to the topic.

Likewise, the device services must use the messaging client (go-mod-messaging for Go services and C alternative for C services) to create a new MessageClient, connect to the message bus and subscribe to configured message topics to receive actuation messages from the command service.

> Because any actuation request may require a response on the part of EdgeX, the command service must provide the means to receive an optional response message from the device services and publish a response back to the 3rd party caller via message bus as well.

Reviewer 3: > Why is the response optional?

No data returned on SET.
Optional on GET. See query parameters here: https://app.swaggerhub.com/apis-docs/EdgeXFoundry1/core-command/2.1.0#/default/get_device_name__name___command_

> Because any actuation request may require a response on the part of EdgeX, the command service must provide the means to receive an optional response message from the device services and publish a response back to the 3rd party caller via message bus as well.

Author: per lenny above.

The command message request messages should contain a correlation identifier which would be added to any response so that the 3rd party system would know to associate the responses to the original request.

The device service message command subscription, response publishing and general handling should be implemented in the SDKs versus individually in the device services.

!!! Note

>     Message responses are not guaranteed (on the part of the device service to the command service or from the command service back to the originating command requester).  Both the command service and the original 3rd party system making the command request bear the burden to manage timeouts (and possibly resending the request) if there is no response.

Reviewer 6: How would this work? Would the command service define a single response timeout setting, or should this be settable on a per-device service, or possibly even on a per-request basis?

If core command times out waiting for a response, would it send its own response indicating a timeout occurred?

Reviewer 3: > If core command times out waiting for a response, would it send its own response indicating a timeout occurred?

IMO, yes send it back to 3rd party requester.

Author: per lenny above.  Added extra comment in this section of the doc.

    

![image](command-msg.png)

### Message Structure

> The messages to the command service and those relayed to the device service would mimic their REST alternatives.  The difference is that the HTTP REST requests

Reviewer 6: ```suggestion
The messages to the command service and those relayed to the device service would mimic their REST alternatives.  The difference is that the HTTP REST requests:
```

> - have a header which contain the correlation id. 

Author: fair enough Lenny.  But in cases where our message bus client is not used, we need to provide an indication of what the message would need to include and look like.  The fact that it is already provided when you use our message bus client is good and helpful.  I mentioned this in a new note.

> - have a header which contain the correlation id. 

Reviewer 3: > But in cases where our message bus client is not used, we need to provide an indication of what the message would need to include and look like..

EdgeX MessageEnvelope is required for all messages on the internal EdgeX MessageBus, whether using our Go or C clients or not.

- have a path that contains the target device and command name

- are made via SET or PUT methods signaling get or set actions on the device.

The message bus and the messages on the bus need this additional information. To address this, the command requests must be embedded in a "wrapper message" structure that would contain the additional elements that would be present in the HTTP request messages.

![image](command-msg-structure.png)

Specifically, the message structure would look something like the following:

``` json
{

> "correlation-id": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",

Author: ??? - @lenny-intel, this is meant to show the general message structure (regardless of whether I am using the message client we have).  Wouldn't the message have to include the correlation id if I originate the message request (as the 3rd party client request)?

> "correlation-id": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",

Reviewer 3: This is in the MessageEnvelope which is required for all message.

"method": "SET",
"deviceName": "sensor01",
"command": "command01",
"request":
    {
        "requestId": "e6e8a2f4-eb14-4649-9e2b-175247911369",
        "AHU-TargetTemperature": "28.5",
        "AHU-TargetBand": "4.0",
        "AHU-TargetHumidity": {
        "Accuracy": "0.2-0.3% RH",
        "Value": 59
    }
}
```
Note, this message structure may not be complete as is meant to represent the general structure.  Additional fields may be added in order to align with the HTTP message structure. 

Response messages would be similarly wrapped so as to preserve the request id - if nothing else.

``` json
{

> "correlation-id": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",

Author: @lenny-intel - The json tag is just mark down indicator.  Why do you think it is not needed?  Is it because it is assumed to be json by default?

> "correlation-id": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",

Reviewer 3: My comment is on line 80, not line 78.

> "response":

Reviewer 6: to be more specific, the response may contain an event with one or more readings

> "response":

Author: The response could contain lots of things but may not necessarily contain an event/reading.  What if it was a SET (aka PUT) call?  Swagger file indicates in that case no event/reading comes back.  So this is just an indication of the general message structure and doesn't represent all types of calls - just the common info.  Or am I missing something?

> "response":

Reviewer 3: IMO, a place hold or comment stating that PUT doesn't return data and GET may have data returned based on the query parameters.

> "response":

Reviewer 3: > The response could contain lots of things but may not necessarily contain an event/reading.

@jpwhitemn, The response will contain nothing or an Event with reading(s). Nothing else is currently in our command responses.

    {
        "requestId": "e6e8a2f4-eb14-4649-9e2b-175247911369",
        "apiVersion": "v2",
        "requestId": "e6e8a2f4-eb14-4649-9e2b-175247911369",
        "statusCode": 0,
        "message": "string"
    }
}
```

### Configuration

The configuration or the command service must provide the details in order to subscribe to command messages from any 3rd party system and publish to a response topic.  The subscription configuration would be similar to how core data is configured to subscribe to messages from any device service and publish to topics to get messages to application services.  However, the command service will actually need to have two sets of subscription and two sets of publishing topics.

> The command service would subscribe to any command topic ("edgex/request/command/#") to receive requests from the 3rd party system.  It must then relay these messages (via publish) to the device service actuation topics.

Reviewer 3: ```suggestion
The command service would subscribe to any command topic ("edgex/command/request/#") to receive requests from the 3rd party system.  It must then relay these messages (via publish) to the device service actuation topics.
```

Author: I was assuming a pattern or request/command/... and response/command/...

@lenny-intel - are you suggesting a reverse pattern of command/request/... and command/response/...?

Reviewer 3: Yes, putting the common static portion of the path first.

The command service must then subscribe to another topic ("edgex/response/actuate/#") to receive actuation responses from the device services.  It then relays those responses on to the 3rd party apps by publishing those responses to a response topic ("edgex/response/command").

![image](command-service-topics.png)

``` toml
[MessageQueue]
Protocol = "redis"
Host = "localhost"
Port = 6379
Type = "redis"
AuthMode = "usernamepassword"  # required for redis messagebus (secure or insecure).
SecretName = "redisdb"
SubscribeResponseActuateTopic = "edgex/response/actuate/#"
PublishActuateTopicPrefix = "edgex/actuate/ " # /<device-service>/<device-name> will be added to this Publish Topic prefix
SubscribeRequestCommandTopic = "edgex/request/command/#"
PublishCommandResponseTopic = "edgex/response/command"
```

Device services must similarly be configured for subscribing to actuation commands and responding to them by publishing to the response topic.

``` toml
[MessageQueue]
Protocol = "redis"
Host = "localhost"
Port = 6379
Type = "redis"
AuthMode = "usernamepassword"  # required for redis messagebus (secure or insecure).
SecretName = "redisdb"

> SubscribeRequestActateTopic = "edgex/request/actuate/<device-service>/#"
> PublishActuateResponseTopic = "edgex/response/actuate/<device-service>/"

Reviewer 3: ```suggestion
SubscribeCommandRequestTopic = "edgex/command/request/<device-service>/#" # /<device-name>/<command-name>/<method>/<request-id> will be added by the publisher
PublishCommandResponseTopic = "edgex/command/response/" # <request-id> will be added to this Publish Topic prefix
```

```

Questions:
- Would we need multiple 3rd party response topics?  Or would one setup separate command services for each 3rd party?
- Do we need separate topics for all the devices or would one on the device service suffice?
  - Ans:  we have defined the deviceName in the payload, so one topic should be sufficient for Device Service- edgex/request/actuate/<device-service>

- Should the command service be allowed to subscribe to the external message broker from a 3rd party? If it subscribes to the EdgeX message bus (internally) only, we might need additional services to relay the data to EdgeX message bus from the external source (depending on security and accessibility to the external message bus).

> - Dynamic configuration is not a user friendly operation. We might want to think about creating additional APIs for Adding/Updating/Deleting/Query the external subscription (and store them to the RedisDB).

Author: Added note for consideration.  I don't think we have to answer/solve this in this iteration.  In the long term, it would seem logical to assume a message bus client would want a message bus based solution to get the list of commands.

> - Is it acceptable for more than one response to be published by the device service on the same correlation ID? Eg, send back "Acknowledged", then "Scheduled", then "Starting", then "Done" statuses?

Reviewer 3: No, Correlation ID life span to from initial requester to response back to requester.

In reverse, response messages from the southside will also be sent through the internal EdgeX message bus to the command service where they can then be bridged to the external message bus and respond to the 3rd party system requester.

!!! Note
    If command did not serve as this external to internal (and vice versa) bridge, then each device service would need to be able to serve as their own external message bus to internal message bus bridge which is inefficient.  Also note that eKuiper is allowed access directly to the internal EdgeX message bus.  This is a special circumstance of 3rd party external system communication as eKuiper is a sister project that is deemed the EdgeX reference implementation rules engine.  In future releases of EdgeX, even eKuiper may be routed through an external to internal message bus bridge for better decoupling and security.

> Any protocol translation or relay service (REST-MQTT, MQTT-REST, REST-REST, MQTT-MQTT) should pass along authentication tokens.

Reviewer 4: Actually, the Kong Auth token does flow through.  MQTT credentials are not part of the MQTT payload.  The intent of this requirement is that we don't want to change the security context of the original request.  When microservices (eventually) start to authenticate requests, we don't want them to reject messages through the bridge because it stripped information about the original caller.  Similarly, we don't want to give the bridge a super-admin token because then the bridge would become a privilege escalation mechanism.  Also, it doesn't seem like a good idea to make the bridge responsible for 1:1 mappings from external credential into internal credentials, as that would require the bridge to be able to fabricate credentials.  That leaves passthrough as the most reasonable choice.

> Any protocol translation or relay service (REST-MQTT, MQTT-REST, REST-REST, MQTT-MQTT) should pass along authentication tokens.

Reviewer 3: @bnevis-i , still doesn't make sense to me for the External MQTT -> Internal MessageBus or HTTP -> Internal MessageBus. We should discuss on next security WG meeting.

> Any protocol translation or relay service (REST-MQTT, MQTT-REST, REST-REST, MQTT-MQTT) should pass along authentication tokens.

Reviewer 4: As discusssed in security WG meeting on 2/2/2022, drop this line for this ADR.  Microservice authentication ADR will be extended to address message-bus based communication in addition to REST.   The current situation only applies to core-command listening to an MQTT bus directly as an alternative interface.  Current proposal is not to have a microservice that generically bridges message busses.

The `envelope` will also contain the API version (something provided in the HTTP path when using REST).

Command requests in HTTP may also contain ds-pushevent and ds-returnevent parameters (for GET commands).  These will be optionally provided key/value pairs represented in the message `envelope`.

``` JSON
{
    "Correlation-ID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",
    "API":"V2",
    "ds-pushevent":"yes",
    "ds-returnevent":"yes",
    ...
}
```

#### Message Payload

The **request** message `payload` to the command service and those relayed to the device service would mimic their HTTP/REST request body alternatives.  The `payload` provides details needed in executing the command at the south side.

In the example GET and PUT messages below, note the `envelope` wraps or encases the message `payload`.  The payload may be empty (as is typical of GET requests).

``` JSON
{
    "Correlation-ID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",
    "API":"v2",
    "ds-pushevent":"yes",
    "ds-returnevent":"yes",
    "payload": 
    {
    }
}

{
    "Correlation-ID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",
    "API":"v2",
    "payload": 
    {

>         "AHU-TargetTemperature": "28.5",

Author: added request id back in.  Cloud had suggested it could be removed.  See his notes earlier in comments.

        "AHU-TargetBand": "4.0",
        "AHU-TargetHumidity": {
            "Accuracy": "0.2-0.3% RH",
            "Value": 59
        }
    }
}

```

The **response** message `payload` would contain the response from the south side, which is typically EdgeX event/reading objects (in the case of GET requests) but would also include any status code, error or service response details.

Example response messages for a GET and PUT request are shown below.  Again, note that the message `envelope` wraps the response `payload`.

``` JSON
{
    "Correlation-ID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",
    "apiVersion": "v2",
    "payload": 
    {
    "requestId": "e6e8a2f4-eb14-4649-9e2b-175247911369",
    "statusCode": 0,
    "message": "string",
    "event": {
        "apiVersion": "string",

>         "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",

Author: took out the duplication.  Will allow the community to debate whether API version needs to be in the response at all.

        "deviceName": "string",
        "profileName": "string",
        "created": 0,
        "origin": 0,
        "readings": [
            "string"
            ],
        "tags": {
            "Gateway-id": "HoustonStore-000123",
            "Latitude": "29.630771",
            "Longitude": "-95.377603"
            }
        }
    }
}

{
    "Correlation-ID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",
    "apiVersion": "v2",
    "payload": 
    {
        "requestId": "e6e8a2f4-eb14-4649-9e2b-175247911369",
        "statusCode": 0,
        "message": "string"
    }
}

```

### Topic Naming

> ![image](command-service-topics.png)

Author: fixed

Example device service configuration is provided below.

``` toml
[MessageQueue]
Protocol = "redis"
Host = "localhost"
Port = 6379
Type = "redis"
AuthMode = "usernamepassword"  # required for redis messagebus (secure or insecure).
SecretName = "redisdb"
SubscribeRequestTopic = "edgex/command/request/#"
PublishResponseTopicPrefix = “edgex/command/response/"  # <device-service>/<device-name>/<command-name>/<method> will be added to this publish topic prefix
```

## Questions

- Do we need separate topics for all the devices or would one on the device service suffice?
    
    - Ans:  we have defined the deviceName in the payload, so one topic should be sufficient for Device Service- edgex/command/request/<device-name>...

- Would clients (non EdgeX services and applications) want to get a list of available commands via message (instead of calling REST)?

    - Ans:  this is a valid question and could be provided via later additions to the command service (or other service like metadata) in the future.  It does not have to be tackled immediately.  

- Dynamic configuration of the message subscription is not a user friendly operation today (requiring configuration changes).

>     - Ans:  We might want to think about creating additional APIs for Adding/Updating/Deleting/Query the external subscription (and store them to the RedisDB).

Author: Added "in the future", but agree that we don't want to go to this level in first implementation.

- Is it acceptable for more than one response to be published by the device service on the same correlation ID? Eg, send back "Acknowledged", then "Scheduled", then "Starting", then "Done" statuses?
    
    - Ans:  No, the correlation id has a life span to/from the initial requester to the response back to the requester.

- Would it make sense to echo the command name into the response, as a reality check?

>     - Ans: solved via topic naming

Reviewer 3: IMO, not needed as we don't do this in the HTTP response. As commened above the repose topic doesn't need the extra path info. The request ID or correlation ID is all that is needed to match the response to the request. No need to make it more complex.

Author: Added Lenny's comment.

- Would sending/receiving binary data (e.g. CBOR) be supported in this north-south message implementation?
    

>     - Ans:  today, command service and device services do not support CBOR operations.  This feature would be explored only when these services support binary/CBOR payloads.

Reviewer 3: Is this correct? I thought they did and there is no reason the payload can't be CBOR encoded. One of the properties on the MessageEnvelope is the content type which indicates how the payload is encoded.

Reviewer 7: CBOR supported on get operations but not set. (although, the c sdk does support set)
Issue: https://github.com/edgexfoundry/device-sdk-go/issues/488

Author: Left as a consideration once we have feature parity on this at the SDK level.

- Use of the message bus communications (by the non-EdgeX 3rd party service or application) would bypass the API Gateway.

>     - Ans:  it would. Pretty much every MQTT broker has a way to do this, where the most common are username/password and certificate-based mutual-auth TLS. SPIFFE/SPIRE is a good way to distribute such certificates if one can manage the requirements for certificate rotation on a continual basis. Otherwise, username/password will probably end up being used (with server-side TLS if the broker is off-box)

Reviewer 3: No longer an issue with the change to Command Service having a External MQTT bridge connection.

In a similar fashion, device services will need to both subscribe and publish to the EdgeX message bus (**internal message bus**) to get command requests and push back any responses to the command service.  Go lang device services will, like the command service, use the go-mod-messaging module and MessagingClient to get command requests and send command responses to and from the EdgeX message bus.  C based device services will use a C alternative to subscribe and publish to the EdgeX message bus (**internal message bus**)

The command service will also need to subscribe to 3rd party MQTT topics (**external message bus**) in order to get command requests from the 3rd party system.  The command service will then relay command requests on to the appropriate device service via the internal message bus (forming the message bus to message bus bridge).  Likewise, the command service will accept responses from the device services on the EdgeX message bus (**internal message bus**) and then publish responses to the 3rd party system via the 3rd party MQTT topics (**external message bus**).

### Message Structure

> In REST based command requests (and responses), the HTTP header is used to pass important information such as the path or target of the request, the HTTP method type (indicating a GET or PUT request), the correlation id and more.  On HTTP responses, the HTTP header provides the information such as the response code (ex: 200 for OK).  The body or payload of the HTTP message contains the request details (such as parameters to a device PUT call) or response information (such as events and associated readings from a GET call).  

Reviewer 6: HTTP headers are used to pass additional information for requests and responses. The path or target of a request is not sent as an HTTP header, it's actually passed as part of the request line. Ex:

`GET /images/logo.png HTTP/1.1
`

Likewise, the HTTP response also is not a header, it's returned as the status line.

`HTTP/1.1 200 OK
`

Example response messages for a GET and PUT request are shown below.  Again, note that the message `envelope` wraps the response `payload`.

``` JSON
{
    "Correlation-ID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",
    "apiVersion": "v2",
    "requestId": "e6e8a2f4-eb14-4649-9e2b-175247911369",
    "statusCode": 0,
    "payload": 
    {
    "message": "string",
    "event": {
        "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
        "deviceName": "string",
        "profileName": "string",
        "created": 0,
        "origin": 0,
        "readings": [
            "string"
            ],
        "tags": {
            "Gateway-id": "HoustonStore-000123",
            "Latitude": "29.630771",
            "Longitude": "-95.377603"
            }
        }
    }
}

{
    "Correlation-ID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",
    "apiVersion": "v2",
    "requestId": "e6e8a2f4-eb14-4649-9e2b-175247911369",
    "statusCode": 0,
    "payload": 
    {
        "message": "string"
    }
}

```

!!! Note

>         Get command responses may include CBOR data.  The CBOR binary object would be enclosed in the payload.

Reviewer 6: I think the point is that the payload is CBOR-encoded if it contains one or more binary readings from a device service. Otherwise it's JSON.

Author: Per Core WG discussion - others felt this was clear as stated.

Author: per the monthly architect's meeting of 3/28/22 - add something to the effect of:  The message envelop (has a content type indicator) will indicate that the payload is either CBOR or JSON.  The same message envelop content type indicator will be used in this message bus communications.

  KeepAlive = "10" # Seconds (must be 2 or greater)
  Retained = "false"
  AutoReconnect = "true"
  ConnectTimeout = "5" # Seconds
  SkipCertVerify = "false" # Only used if Cert/Key file or Cert/Key PEMblock are specified

## new configuration to allow device services to also communicate via message bus with core command
CommandRequestTopic = "edgex/command/request/#"         # subscribing for inbound command requests
CommandResponseTopicPrefix = "edgex/command/response/"  # publishing outbound command responses; <device-service>/<device-name>/<command-name>/<method> will be added to this publish topic prefix
```

!!! Note
        Most of the device service configuration is existing based on its need to already communicate with the message bus for publishing events/readings.  The last two lines are added to allow device services to subscribe and publish command messages from/to the message bus.

## Questions

- Do we need separate topics for all the devices or would one on the device service suffice?
    
    - Ans:  we have defined the deviceName in the parameterized topic, so one topic should be sufficient for Device Service- edgex/command/request/<device-name>...

> - Would clients (non EdgeX services and applications) want to get a list of available commands via message (instead of calling REST)?

Reviewer 6: I would think it's important that we provide parity with the existing REST implementation. If we don't provide a means to get the command list via message bus, then we force 3rd party applications to use both REST and message bus.

Author: Added sections to cover queries via message bus

# North-South Messaging

## Status

**Proposed**
(as of 12/21/21)

## Context and Proposed Design

Today, data flowing from sensors/devices (the “southside”) through EdgeX to enterprise applications, databases and cloud-based systems (the “northside”) can be accomplished via REST or Message bus.  That is, sensor or device data collected by a device service can be sent via REST or message bus to core data.  Core data then relays the data to application services via message bus, but the sensor data can also be sent directly from device services to application services via message bus (bypassing core data).  The message bus is implemented via Redis Pub/Sub (default) or via MQTT.  From the application services, data can be sent to northside endpoints in any number of ways – including via MQTT.

So, in summary, data can be collected from a sensor or device and be sent from the southside to the northside entirely using message bus technology when desired.

> !!! Note

Author: agreed to remove per monthly architect's meeting of 3/28/22

    The message bus may also be implemented in Zero MQ but with some limitations.  As Zero MQ is being deprecated in EdgeX, it will not be considered further in this ADR.

Today, communications from a 3rd party system (enterprise application, cloud application, etc.) to EdgeX in order to acuate a device or get the latest information from a sensor is accomplished via REST.  The 3rd party system makes a REST call of the command service which then relays a request to a device service also using REST.  There is no built in means to make a message-based request of EdgeX or the devices/sensors it manages.  Note, these REST calls are optionally made via the API Gateway in order to provide access control.

> In a future release of EdgeX, there is a desire to allow 3rd party systems to make requests of the southside via message bus.  Specifically, a 3rd party system will send a command request via message to the command service via the allowed message bus implementations (which could be MQTT or Redis Pub/Sub today).  The command service would then relay the message request via message bus to the managing device service.  (The command service would also perform any translations on the request as it does for REST requests today.) The device service would use the message to trigger action on the device/sensor as it does when it receives a REST request today and respond via message bus back to the command service.  In turn, the command service would respond to the 3rd party system via message bus.

Author: per monthly architect's meeting 3/28/22 - remove for clarity.

![image](command-message-bridge.png)

!!! Note
    For the purposes of this initial north-to-south message bus communications, external 3rd party communications to the command service will be **limited to use of MQTT**.

### Core Command as Message Bus Bridge

The core command service will serve as the EdgeX entry point for external, north-to-south message bus requests to the south side.

![image](command-msg.png)

3rd party systems should not be granted access to the EdgeX internal message bus.  Therefore, in order to implement north to south communications via message bus (specifically MQTT), the command service needs to take messages from the 3rd party or external MQTT topics and pass them internally onto the EdgeX internal message bus where they can eventually be routed to the device services and then on to the devices/sensors (southside).

In reverse, response messages from the southside will also be sent through the internal EdgeX message bus to the command service where they can then be bridged to the external MQTT topics and respond to the 3rd party system requester.

!!! Note

>     If command did not serve as this external to internal (and vice versa) bridge, then each device service would need to be able to serve as their own external MQTT topic to internal message bus bridge which is inefficient.  Also note that eKuiper is allowed access directly to the internal EdgeX message bus.  This is a special circumstance of 3rd party external system communication as eKuiper is a sister project that is deemed the EdgeX reference implementation rules engine.  In future releases of EdgeX, even eKuiper may be routed through an external to internal message bus bridge for better decoupling and security.

Author: per monthly architect's meeting of 3/28/22, will pull first sentence

### Message Bus Subscriptions and Publishing

The command service will require the means to publish messages to device services via the EdgeX message bus (**internal message bus**).  It would use the messaging client (go-mod-messaging) to create a new MessageClient, connect to the message bus, and publish to designated request message topics (see topic configuration below).

The command service will also need to subscribe to the EdgeX message bus (**internal message bus**) in order to receive responses from the device services after a request by message bus has been made.  Again, core command will use the go-mod-messaging MessageClient to subscribe and receive response messages from the device services.

> In a similar fashion, device services will need to both subscribe and publish to the EdgeX message bus (**internal message bus**) to get command requests and push back any responses to the command service.  Go lang device services will, like the command service, use the go-mod-messaging module and MessagingClient to get command requests and send command responses to and from the EdgeX message bus.  C based device services will use a C alternative to subscribe and publish to the EdgeX message bus (**internal message bus**)

Author: added

> In a similar fashion, device services will need to both subscribe and publish to the EdgeX message bus (**internal message bus**) to get command requests and push back any responses to the command service.  Go lang device services will, like the command service, use the go-mod-messaging module and MessagingClient to get command requests and send command responses to and from the EdgeX message bus.  C based device services will use a C alternative to subscribe and publish to the EdgeX message bus (**internal message bus**)

Author: added

> The command service will also need to subscribe to 3rd party MQTT topics (**external message bus**) in order to get command requests from the 3rd party system.  The command service will then relay command requests on to the appropriate device service via the internal message bus (forming the message bus to message bus bridge).  Likewise, the command service will accept responses from the device services on the EdgeX message bus (**internal message bus**) and then publish responses to the 3rd party system via the 3rd party MQTT topics (**external message bus**).

Reviewer 3: What about request from App Service via the internal message bus?

Author: @lenny-intel - see my additions to the latest under "EdgeX Service (Internal) Message Bus Requests"

In the case of a command query, the REST queries would return REST endpoints to call in which to make the query.  For example, the REST query would return core command paths and urls and parameters to use to make REST command requests (as shown in the example below).

``` JSON
 "coreCommands": [
        {
          "name": "coolingpoint1",
          "get": true,
          "path": "/api/v2/device/name/testDevice1/command/coolingpoint1",
          "url": "http://localhost:59882",
          "parameters": [
            {
              "resourceName": "resource1",
              "valueType": "Int32"
            }
          ]
        }
 ]
```

When using messaging to make the "queries" the response message must return information about how to pass a message to the appropriate topic to make the command request.  Therefore, the query response when using messaging might include something like the following:

``` JSON
 "coreCommands": [
        {
          "name": "coolingpoint1",
          "get": true,

>           "topic": "/my-app/command/request/testDevice1/coolingpoint1/set",

Reviewer 3: @tonyespy,  As described below the `<method>` is part of the topic . 
@jpwhitemn , here we need to be able to handle commands that can be both get & set. To handle this we need to:
1.  drop the get/set Boolean and have getTopic and setTopic.
2. drop the `/set` or `/get` from the topic and the user has to add them based on the get/set Boolean (similar to rest where the use has to select GET or POST).

>           "topic": "/my-app/command/request/testDevice1/coolingpoint1/set",

Author: fixed

              "valueType": "Int32"
            }          ]
        }
 ]
```

!!! Note
        If desired, the query commands could return information to make *either* a REST or message request.  Presumably, the query responses would be the same then for both REST and message query requests so that information returned allows the 3rd party application to choose whether to use REST or message bus to make the command requests. 

### Message Structure

In REST based command requests (and responses), the HTTP request line contains important information such as the path or target of the request, and the HTTP method type (indicating a GET or PUT request).  The HTTP status line provides the information such as the response code (ex: 200 for OK).  The body or payload of the HTTP message contains the request details (such as parameters to a device PUT call) or response information (such as events and associated readings from a GET call).  

In a message bus world, there is no HTTP header to carry request or response details.  Therefore, the message must be organized into an `envelope` and `payload` section to serve just like the HTTP header and body do in HTTP request/response messages.

> The message topic names will, in many ways, serve as the HTTP paths and methods do in helping to specify the device receiver of any command request (see topic names below)

Author: written to hopefully make sense.

> The message topic names will, in many ways, serve as the HTTP paths and methods do in helping to specify the device receiver of any command request (see topic names below)

Author: written to hopefully make sense.

![image](command-msg-structure.png)

#### Message Envelope

> Messages are just JSON text.  The outer most JSON object represents the message `envelope` for command request and response messages.  The `envelope` will contain a correlation identifier property which will be added to any relayed request message as well as the response message envelope so that the 3rd party system will know to associate the responses to the original request.

Author: I had this explanation in the question/answer area but created a special note with this in it to provide more clarity.

> Messages are just JSON text.  The outer most JSON object represents the message `envelope` for command request and response messages.  The `envelope` will contain a correlation identifier property which will be added to any relayed request message as well as the response message envelope so that the 3rd party system will know to associate the responses to the original request.

Author: I had this explanation in the question/answer area but created a special note with this in it to provide more clarity.

The `envelope` will also contain the API version (something provided in the HTTP path when using REST).

> Command requests in HTTP may also contain ds-pushevent and ds-returnevent query parameters (for GET commands).  These will be optionally provided key/value pairs represented in the message `envelope`'s query parameters (and optionally allows for other parameters in the future).

Reviewer 3: @tonyespy Core Command doesn't support query parameters for PUT
https://app.swaggerhub.com/apis-docs/EdgeXFoundry1/core-command/2.1.0#/default/put_device_name__name___command_

``` JSON
{

>     "Correlation-ID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",

Author: Added this issue to the EdgeX 3.0 project board https://github.com/orgs/edgexfoundry/projects/48#card-80220387 but did not alter the ADR at this time to keep things consistent.

>     "Correlation-ID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",

Author: Added this issue to the EdgeX 3.0 project board https://github.com/orgs/edgexfoundry/projects/48#card-80220387 but did not alter the ADR at this time to keep things consistent.

    "API":"V2",
    "queryParams": {

>         "ds-pushevent":"yes",

Reviewer 3: @tonyespy , yes preferably but this is how they are in REST currently, which needs to be fixed in 3.0. IMO they should be bool here so we do it correct and don't have to fix in 3.0

>         "ds-pushevent":"yes",

Author: Also added to EdgeX 3.0 board https://github.com/orgs/edgexfoundry/projects/48#card-80220476 but made no changes to ADR for consistency.

>         "ds-pushevent":"yes",

Author: Also added to EdgeX 3.0 board https://github.com/orgs/edgexfoundry/projects/48#card-80220476 but made no changes to ADR for consistency.

        "ds-returnevent":"yes",
     }
     ...
}
```

#### Command Message Payload

The **request** message `payload` to the command service and those relayed to the device service would mimic their HTTP/REST request body alternatives.  The `payload` provides details needed in executing the command at the south side.

> In the example GET and PUT messages below, note the `envelope` wraps or encases the message `payload`.  The payload may be empty (as is typical of GET requests).

Reviewer 3: @tonyespy , that is an implementation detail and can be left our in the JSON. But for here it illustrates that the payload is empty.

    - Per core WG meeting discussion of 2/24/22 - do we really need the request id (is it redundant based on already having correlation id)?
    - If we have request id, should it be in the payload?
    - Per core WG meeting discussion of 2/24/22 - do we have status code?  If so should it mimic the REST/HTTP status code responses?  Do we really want to mimic HTTP in our message bus approach?  As suggested by @farshidtz, maybe we should just have an `error` boolean and then have the message indicate the error condition.  
    - If we have a status code or error code, where does it belong?  In the payload or above it (as it would be in the header in REST)?  As a reference, the [IoTAAP MQTT to REST bridge](https://docs.iotaap.io/docs-rest/) provides a status code to message string translation as an example means to handle this problem.  Should we use something similar?

#### Query Message Payload

The **request** message `payload` to query the command service would mimic their HTTP/REST request body alternatives.  The `payload` provides details needed in executing the command at the south side.

> In the example query to get all commands below, note the `envelope` wraps or encases the message `payload`.  The payload will be empty.  The query parameters will include the offset and limit (as per the REST counter parts).

Reviewer 3: @tonyespy , this providing same functionality as Core Command REST API here, no Device Services.
https://app.swaggerhub.com/apis-docs/EdgeXFoundry1/core-command/2.1.0#/default/get_device_name__name_
and here:
https://app.swaggerhub.com/apis-docs/EdgeXFoundry1/core-command/2.1.0#/default/get_device_all

``` JSON
{
    "Correlation-ID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",
    "API":"v2",
    "requestId": "e6e8a2f4-eb14-4649-9e2b-175247911369",
    "queryParams": {
        "offset":0,
        "limit":20,
     }
    "payload": 
    {
    }
}

In the example query to get commands for a specific device by name, the device name is the lone query parameter (again mirroring the REST interface).

{
    "Correlation-ID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",
    "API":"v2",
    "requestId": "e6e8a2f4-eb14-4649-9e2b-175247911369",
    "queryParams": {

>         "name":"coolingpoint1"

Reviewer 3: @jpwhitemn , IMO the `deviceName` should be part of the topic similar to how it is part of the REST endpoint path.

     }
    "payload": 
    {
    }
}

```

The **response** message `payload` for queries would contain the information necessary to make a message-base command request.

An example response message is shown below.  Again, note that the message `envelope` wraps the response `payload`.

``` JSON
{
    "Correlation-ID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",
    "apiVersion": "v2",
    "requestId": "e6e8a2f4-eb14-4649-9e2b-175247911369",
    "statusCode": 0,

>     "totalCount": 2,

Author: removed

>     "totalCount": 2,

Author: removed

                "topic": "/my-app/command/request/testDevice1/coolingpoint1/get",
                "url": "broker.address:1883",
                "parameters": [
                    {
                    "resourceName": "resource1",
                    "valueType": "Int32"
                    }
                ]
                }
            ]
            },
            {
            "deviceName": "testDevice1",
            "profileName": "testProfile",
            "coreCommands": [
                {
                "name": "coolingpoint1",
                "set": true,
                "topic": "/my-app/command/request/testDevice1/coolingpoint1/set",
                "url": "broker.address:1883",
                "parameters": [
                    {
                    "resourceName": "resource5",
                    "valueType": "String"
                    },
                    {
                    "resourceName": "resource6",
                    "valueType": "Bool"
                    }
                ]
                }
            ]
            }
        ]
    }
}

```

### Topic Naming

![image](command-service-topics.png)

#### 3rd party system topics

> The 3rd party system or application must publish command requests messages to its own MQTT topic (**external message bus**) and subscribe to responses from the same.  Messages topics should follow a standard such as (where `my-app` is replaced by the 3rd party application or system name):

Reviewer 3: Agreed, 3rd party apps are going to have to conform to the Edgex topic format.

- Publishing command request topic: `/my-app/command/request/<device-name>/<command-name>/<method>`
- Subscribing command response topic: `/my-app/command/response/#`

For queries, the following topics are used
- Publishing query command request topic: `/my-app/commandquery/request`
- Subscribing query command response topic: `/my-app/commandquery/response`

!!! Note

>     Because EdgeX can suggest but not dictate the naming standard for 3rd party MQTT topics, these names are representative for clarity, but not required.  The suffix inclusion of <device-name>/<command-name>/<method> is required for the command service to be able to understand and direct the request to the appropriate device/sensor through the correct device service.  Alternately, this information could be placed in the message header and the topic names made generic, but this is different than what we have done for south to north communications.

Author: per monthly architect's meeting of 3/28/22, pull this note.  This is the same for events today.

#### command service configuration

Example command service configuration is provided below.

``` toml
[MessageQueue]
    [InternalMessageQueue]
    Protocol = "redis"
    Host = "localhost"
    Port = 6379
    Type = "redis"
    RequestTopicPrefix = "edgex/command/request/"   # for publishing requests to the device service; <device-service>/<device-name>/<command-name>/<method> will be added to this publish topic prefix
    ResponseTopic = "edgex/command/response/#”      # for subscribing to device service responses
    AuthMode = "usernamepassword"                   # required for redis messagebus (secure or insecure).
    SecretName = "redisdb"
    [ExternalMessageQueue]
    Protocol = "tcp"
    Host = "localhost"
    Port = 6378
    Type = "mqtt"
    RequestTopic = "my-app/command/request/#"           # for subscribing to 3rd party command requests
    ResponseTopicPrefix = "my-app/command/response/"    # for publishing responses back to 3rd party systems /<device-name>/<command-name>/<method> will be added to this publish topic prefix
    [ExternalQueryQueue]

>     Protocol = "tcp"

Author: per monthly architect's meeting of 3/28/22 - collapse the two external query queue sections into on.  come up with different names for topic and prefix for each

```

!!! Note
        Most of the device service configuration is existing based on its need to already communicate with the message bus for publishing events/readings.  The last two lines are added to allow device services to subscribe and publish command messages from/to the message bus.

## Questions

- Do we need separate topics for all the devices or would one on the device service suffice?
    
    - Ans:  we have defined the deviceName in the parameterized topic, so one topic should be sufficient for Device Service- edgex/command/request/<device-name>...

- Would clients (non EdgeX services and applications) want to get a list of available commands via message (instead of calling REST)?

    - Ans:  this is a valid question and could be provided via later additions to the command service (or other service like metadata) in the future.  It does not have to be tackled immediately.  

- Dynamic configuration of the message subscription is not a user friendly operation today (requiring configuration changes).

>     - Ans:  In the future, we might want to think about creating additional APIs for Adding/Updating/Deleting/Query the external subscription (and store them to the RedisDB).

Author: per monthly architect's meeting of 3/28/22 - add note about using Consul as alternative

    
    - Ans:  No, the correlation id has a life span to/from the initial requester to the response back to the requester.

- Would it make sense to echo the command name into the response, as a reality check?
    - Ans: solved via topic naming.  Also, per @lenny-intel: "not needed as we don't do this in the HTTP response. The response topic doesn't need the extra path info. The request ID or correlation ID is all that is needed to match the response to the request. No need to make it more complex."

- Would sending/receiving binary data (e.g. CBOR) be supported in this north-south message implementation?
    
    - Ans:  today, command service and device services support CBOR get operations but not set (C SDK suppports both).  [Suggest getting feature parity](https://github.com/edgexfoundry/device-sdk-go/issues/488) in place between the SDKs before exploring CBOR support messaging binary/CBOR payloads.
    - Ans update per Monthly Architect's meeting of 2/28/22: support get with CBOR payload.

- Use of the message bus communications (by the non-EdgeX 3rd party service or application) would bypass the API Gateway.

>     - Ans:  not an issue since the command service is serving as external to internal message bus broker.

Author: per monthly architect's meeting of 3/28/22, I will add this comment as stated by @tonyespy.

> - Note a number of open questions in the Message Structure section that still need to be addressed.

Reviewer 6: This ^^

Author: addressed in monthly architect's meeting of 3/28/22

Similar to EdgeX application services, the command service will have access to an external MQTT broker to get command requests and send 3rd parties a response.  This will require the command service to have two message queue configuration settings (internal and external).

#### command service configuration

Example command service configuration is provided below.

``` toml
[MessageQueue]
    [InternalMessageQueue]
    Protocol = "redis"
    Host = "localhost"
    Port = 6379
    Type = "redis"
    RequestTopicPrefix = "edgex/command/request/"   # for publishing requests to the device service; <device-service>/<device-name>/<command-name>/<method> will be added to this publish topic prefix
    ResponseTopic = "edgex/command/response/#”      # for subscribing to device service responses
    AuthMode = "usernamepassword"                   # required for redis messagebus (secure or insecure).
    SecretName = "redisdb"

>     [ExternalMessageQueue]

Reviewer 3: ```suggestion
    [ExternalMQTT]
```

    Protocol = "tcp"
    Host = "localhost"

>     Port = 6378

Reviewer 3: ```suggestion
    Port = 1883
```

>     Type = "mqtt"

Reviewer 3: Remove since MQTT is only type supported

In the case of command queries, the REST responses include the actual REST command endpoints.  For example, the REST query would return core command paths, urls and parameters used to construct REST command requests (as shown in the example below).

``` JSON
 "coreCommands": [
        {
          "name": "coolingpoint1",
          "get": true,
          "path": "/api/v2/device/name/testDevice1/command/coolingpoint1",
          "url": "http://localhost:59882",
          "parameters": [
            {
              "resourceName": "resource1",
              "valueType": "Int32"
            }
          ]
        }
 ]
```

When using messaging to make the "queries" the response message must return information about how to pass a message to the appropriate topic to make the command request.  Therefore, the query response when using messaging would include something like the following:

``` JSON
 "coreCommands": [
        {
          "name": "coolingpoint1",
          "topic": "/edgex/command/request/testDevice1/coolingpoint/get",

>           "url": "broker.address:1883",

Author: pulled URL and added a note.

!!! Note
        A Correlation ID (see this [article](https://www.rapid7.com/blog/post/2016/12/23/the-value-of-correlation-ids/) for a more detailed description) is a unique value that is added to every request and response involved in a transaction which could include multiple requests/responses between one or more microservices. It's not meant to correlate requests to responses, its meant to label every message involved in a potentially multi-request transaction.

        A Request ID should be an identifier returned on the response to a request (providing traceability between single request/response).

The `envelope` will also contain the API version (something provided in the HTTP path when using REST).

Command requests in HTTP may also contain ds-pushevent and ds-returnevent query parameters (for GET commands).  These will be optionally provided key/value pairs represented in the message `envelope`'s query parameters (and optionally allows for other parameters in the future).

``` JSON
{
    "Correlation-ID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",
    "API":"V2",
    "queryParams": {
        "ds-pushevent":"yes",

>         "ds-returnevent":"yes",

Author: According to the Swagger docs for REST requests, a response is still provided, but with no events.  I would assume we want the same behavior.  Therefore, yes, an envelope but with no payload.  Added a note in this section to clarify.

     }
     ...
}
```

#### Command Message Payload

The **request** message `payload` to the command service and those relayed to the device service would mimic their HTTP/REST request body alternatives.  The `payload` provides details needed in executing the command at the south side.

> In the example GET and PUT messages below, note the `envelope` wraps or encases the message `payload`.  The payload may be empty (as is typical of GET requests).

Reviewer 6: Why include the `payload` if it's empty? Why not just make it optional element?

Author: pulled the payload and added a note that it was optional.

Example response messages for a GET and PUT request are shown below.  Again, note that the message `envelope` wraps the response `payload`.

``` JSON
{
    "Correlation-ID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",
    "apiVersion": "v2",
    "requestId": "e6e8a2f4-eb14-4649-9e2b-175247911369",
    "errorCode": 0,
    "payload": 
    {
    "event": {
        "apiVersion": "v2",
        "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
        "deviceName": "string",
        "profileName": "string",
        "created": 0,
        "origin": 0,
        "readings": [
            "string"
            ],
        "tags": {
            "Gateway-id": "HoustonStore-000123",
            "Latitude": "29.630771",
            "Longitude": "-95.377603"
            }
        }
    }
}

{
    "Correlation-ID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",
    "apiVersion": "v2",
    "requestId": "e6e8a2f4-eb14-4649-9e2b-175247911369",
    "errorCode": 1,
    "payload": 
    {
        "message": "string"
    }
}

> ```

Reviewer 6: Note, the use of pre-formatted text makes it a bit of a pita to review due to the horizontal scrolling involved. Just a suggestion to maybe not use this in future ADRs.

Author: noted

    - Per core WG meeting discussion of 2/24/22 - do we really need the request id (is it redundant based on already having correlation id)?
        - Ans (per monthly architect's meeting of 3/28/22): request id and correlation id are two different things.  Request id should be returned on the response to a request (providing traceability between single request/response).  Correlation id is used to track across an entire transaction of many service request/responses (providing traceability across a "transaction" which is often across many services).  So both are needed and should be kept.
    - If we have request id, should it be in the payload?
        - Ans (per monthly architect's meeting of 3/28/22): This should be kept in the message envelope (not the message payaload).
    - Per core WG meeting discussion of 2/24/22 - do we have status code?  If so should it mimic the REST/HTTP status code responses?  Do we really want to mimic HTTP in our message bus approach?  As suggested by @farshidtz, maybe we should just have an `error` boolean and then have the message indicate the error condition. 

>         - Ans (per monthly architect's meeting of 3/28/22): We will use errorCode to provide the indication of error.  The errorCode will be 0 (no error) or 1 (indicating error) as the two enums for error conditions today.  In the future, if we determine the need for additional errors, we will add more "enums" or error code numbers.  When there is an error (with errorCode set to 1), then the payload contains a message string indicating more information about the error.  When there is no error (errorCode 0) then there is no message string in the payload. 

Reviewer 6: I think this should be re-visited, as we already have enums for REST for things like:

- 400 - request is in an invalid state (note, this probably needs re-wording in our swagger documentation as I think it really means invalid request)
- 404 - resource doesn't exist
- 423 - the device is locked or down
- 500 - internal error

See: https://app.swaggerhub.com/apis-docs/EdgeXFoundry1/core-command/2.1.0#/default/get_device_name__name___command_

Author: This was covered in at least one meeting before and the thoughts were that we didn't need the complexity of all the HTTP responses.  I'll review this issue in the 4/27 TSC.

Author: Implementation experimentation to be done - during implementation, some options to be explored and decided on PR review.  This issue should not hold up the approval of the ADR.  Per TSC meeting of 4/27

Author: Added note to the ADR that this issue will be addressed/finalized during implementation.

    - If we have a status code or error code, where does it belong?  In the payload or in the envelope (as it would be in the header in REST)?  As a reference, the [IoTAAP MQTT to REST bridge](https://docs.iotaap.io/docs-rest/) provides a status code to message string translation as an example means to handle this problem.  Should we use something similar?
        - Ans (per monthly architect's meeting of 3/28/22): errorCode should be in the message envelope not the payload.  When there is an error, the payload contains a single message string.

#### Query Message Payload

The **request** message `payload` to query the command service would mimic their HTTP/REST request body alternatives.  The `payload` provides details needed in executing the command at the south side.

> In the example query to get all commands below, note the `envelope` wraps or encases the message `payload`.  The payload will be empty.  The query parameters will include the offset and limit (as per the REST counter parts).

Reviewer 6: See previous comment about empty payloads, just don't include it...

``` JSON
{
    "Correlation-ID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",
    "apiVersion": "v2",
    "requestId": "e6e8a2f4-eb14-4649-9e2b-175247911369",
    "queryParams": {
        "offset":0,
        "limit":20,
     }
    "payload": 
    {
    }
}

In the example query to get commands for a specific device by name, the device name would be in the topic, so the query message would be without information.

{
    "Correlation-ID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",
    "apiVersion": "v2",
    "requestId": "e6e8a2f4-eb14-4649-9e2b-175247911369",

>     "queryParams": {

Reviewer 6: Same as with empty payload, just don't include `queryParams` if there aren't any...

Author: removed

                "topic": "/edgex/command/request/testDevice1/coolingpoint1/get",
                "url": "broker.address:1883",
                "parameters": [
                    {
                    "resourceName": "resource1",
                    "valueType": "Int32"
                    }
                ]
                }
            ]
            },
            {
            "deviceName": "testDevice1",
            "profileName": "testProfile",
            "coreCommands": [
                {
                "name": "coolingpoint1",
                "set": true,
                "topic": "/edgex/command/request/testDevice1/coolingpoint1/set",
                "url": "broker.address:1883",
                "parameters": [
                    {
                    "resourceName": "resource5",
                    "valueType": "String"
                    },
                    {
                    "resourceName": "resource6",
                    "valueType": "Bool"
                    }
                ]
                }
            ]
            }
        ]
    }
}

```

### Topic Naming

![image](command-service-topics.png)

#### 3rd party system topics

> The 3rd party system or application must publish command requests messages to its own MQTT topic (**external message bus**) and subscribe to responses from the same.  Messages topics should follow a standard such as:

Reviewer 6: I thought we discussed this and said that its not really up to the 3rd party system to define these topics? It's our mechanism, so we should define the topics the 3rd party applications should use when they want to use Core Command.

Author: overlooked text that was not updated. fixed.

#### command service topics

The command service must subscribe to the request topics of the 3rd party MQTT topic (**external message bus**) to get command requests, publish those to a topic to send them to a device service via the EdgeX message bus (**internal message bus**), subscribe to response messages on topics from device services (**internal**), and then publish response messages to a topic on the 3rd party MQTT broker (**external**).  Message topics for the command service would follow the following standard:

- Subscribing to 3rd party command request topics: `edgex/command/request/#`
- Publishing to device service request topic: `edgex/command/request/<device-service>/<device-name>/<command-name>/<method>`
- Subscribing to device service command response topics: `edgex/command/response/#`
- Publishing to 3rd party command response topic: `edgex/command/response/<device-name>/<command-name>/<method>`

For queries, the following topics are used:

- Subscribing to 3rd party command query request topics: `edgex/commandquery/request`
- Publishing to 3rd party command query response topic: `edgex/commandquery/response`

!!! Note

>     Because EdgeX can suggest but not dictate the naming standard for 3rd party MQTT topics, the 3rd party topic names are representative for clarity.

Reviewer 6: See previous comment

Author: pulled

## EdgeX Service (Internal) Message Bus Requests

Application services (or other EdgeX services in the future) may want to also use message communications to make command requests.  Application services make command requests today via REST.

In order to support this, the following need to be added:

- The command service will also need an internal request topic and internal response topic prefix configuration to allow internal EdgeX services to make command requests (and query requests).

    ``` toml
        [MessageQueue]
            [InternalMessageQueue]
            Protocol = "redis"
            Host = "localhost"
            Port = 6379
            Type = "redis"
            RequestTopicPrefix = "edgex/command/request/"   # for publishing requests to the device service; <device-service>/<device-name>/<command-name>/<method> will be added to this publish topic prefix
            ResponseTopic = "edgex/command/response/#”      # for subscribing to device service responses

>             InternalRequestCommandTopic = "/command/request/#"     # for subscribing to internal command requests

Reviewer 6: Do the internal topics really need to be configurable?

Author: I guess they wouldn't have to be, but then again neither would the other topics we are using.  Will bring this up at TSC on 4/27.

    ``` TOML
        [MessageQueue]
            [InternalMessageQueue]
            Protocol = "redis"
            Host = "localhost"
            Port = 6379
            Type = "redis"
            RequestTopicPrefix = "edgex/command/request/"   # for publishing requests to the device service; <device-service>/<device-name>/<command-name>/<method> will be added to this publish topic prefix
            ResponseTopic = "edgex/command/response/#”      # for subscribing to device service responses
            InternalRequestCommandTopic = "/command/request/#"     # for subscribing to internal command requests
            InternalResponseCommandTopicPrefix = "/command/response/"    # for publishing responses back to internal service /<device-name>/<command-name>/<method> will be added to this publish topic prefix
            InternalRequestQueryTopic = "/commandquery/request"
            InternalResponseQueryTopic = "/commandquery/response"
            AuthMode = "usernamepassword"                   # required for redis messagebus (secure or insecure).
            SecretName = "redisdb"
    ```

> - A new command message client will need to be created to allow internal services (app services in this instance) to conveniently use the message bus communications with core command.

Reviewer 3: Need to mention the Clients configuration will also be expanded to include corresponding topics and `UseMessageBus` flag enabling the new messaging based CommandClient to be created. Something like:
```
[Clients]
  [Clients.core-command]
  UseMessageBus = true
  Protocol = "redis"
  Host = "localhost"
  Port = 6379
  CommandRequestTopicPrefix = "/command/request"  /<device-name>/<command-name>/<method> will be added to this publish topic prefix
  CommandResponseTopic = "/command/response/#"
  CommandQueryRequestTopic = "/commandquery/request"
  CommandQueryResponseTopic = "/commandquery/response"
```
